### PR TITLE
[WIP] Reimplements RuntimeEnvService in raylet.

### PR DESCRIPTION
**Pull request by rynewang**
### Mission Statement
We have python grpc services and clients for runtime env agent, and we want to get rid of both of them because we don't want ray depend on grpc because the latter causes some version conflicts with the customer code.

### The plan

- [x] Adds a C++ service interface for RuntimeEnvService.
- [ ] Implement it by spawning shell -> python -> plugins.
- [x] Initiates the C++ service in NodeManager behind a feature flag.
- [ ] (some time later than this PR) remove the feature flag and the old AgentManager's ability to talk to Python side.
- [ ] Remove grpcio dependency in `ray` mininmal.
## Related issue number

This is for #35472.

## Checks

- [ ] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a 
           method in Tune, I've added it in `doc/source/tune/api/` under the 
           corresponding `.rst` file.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: src/ray/raylet/agent_manager.h [hunk 1]
```diff
@@ -89,6 +79,10 @@ class AgentManager : public rpc::AgentManagerServiceHandler {
   virtual void DeleteRuntimeEnvIfPossible(const std::string &serialized_runtime_env,
                                           DeleteRuntimeEnvIfPossibleCallback callback);
 
+  // Wraps this AgentManager to a RuntimeEnvManagerClient. The caller must ensure the
+  // Client lives shorter than this AgentManager.
```
Comment (pcmoritz) [inline src/ray/raylet/agent_manager.h:83]: Hmm, I think you should have the `AgentManagerAsRuntimeEnvManagerClient` hold a smart pointer to `AgentManager` and lift this restriction on the lifetime. It removes a source of problems and there are no efficiency gains to be had in making it a plain pointer :)

**Comment by pcmoritz** [comment]
So the PR mostly makes sense to me and I agree with the overall strategy. In particular, introducing `AgentManagerAsRuntimeEnvManagerClient` for the migration makes sense to me.

There is some extra complexity I don't understand though:

Why are you modifying `AgentManager::GetOrCreateRuntimeEnv` and `AgentManager::DeleteRuntimeEnvIfPossible` at all? You can just leave that code unmodified and then you also won't need the `ThinWrapperRuntimeEnvManagerClient`, right?

So basically, intercept the old agent_manager by replacing it with RuntimeEnvManagerClient -- if the feature flag is off, you use the old implementation unmodified through the AgentManagerAsRuntimeEnvManagerClient and if the feature flag is on, you just don't use the AgentManager at all and instead use your new implementation that is in the raylet.

**Comment by rynewang** [comment]
I was trying to extract out common code useful to do `rpc <-> c++ method` translations into the ThinWrapper and it can be useed by both AgentManager and the C++ server's client. However now that we decided to remove AgentManager at all, this is no longer the case. Removing code...

File: src/ray/raylet/agent_manager.h (hunk 1 above)
Comment (rynewang) [reply src/ray/raylet/agent_manager.h:83]: ditto, now we are removing agent manager so it is no longer needed.

**Comment by SongGuyang** [comment]
We already started refactoring the runtime env agent https://github.com/ray-project/ray/pull/35292 and I think the two works are conflict.
